File: common/src/main/java/com/pluxity/facility/path/FacilityPathService.java
package com.pluxity.facility.path;

import static com.pluxity.global.constant.ErrorCode.NOT_FOUND_FACILITY_PATH;

import com.pluxity.facility.Facility;
import com.pluxity.global.exception.CustomException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.StringUtils;

@Service
@RequiredArgsConstructor
public class FacilityPathService {
    private final FacilityPathRepository facilityPathRepository;

    @Transactional(readOnly = true)
    public FacilityPath findById(Long id) {
        return facilityPathRepository
                .findById(id)
                .orElseThrow(() -> new CustomException(NOT_FOUND_FACILITY_PATH, id));
    }

    @Transactional
    public void update(Long pathId, String name, String type, String path) {
        FacilityPath facilityPath = findById(pathId);
        if (StringUtils.hasText(name)) {
            facilityPath.updateName(name);
        }
        if (StringUtils.hasText(type)) {
            facilityPath.updatePathType(PathType.from(type));
        }
        if (StringUtils.hasText(path)) {
            facilityPath.updatePath(path);
        }
    }

    @Transactional
    public void save(Facility facility, String name, String type, String path) {
        facilityPathRepository.save(
                FacilityPath.builder()
                        .facility(facility)
                        .name(name)
                        .pathType(PathType.valueOf(type))
                        .path(path)
                        .build());
    }

    @Transactional
    public void delete(Long pathId) {
        facilityPathRepository.delete(findById(pathId));
    }
}

File: common/src/main/java/com/pluxity/facility/path/FacilityPathService.kt
package com.pluxity.facility.path

import com.pluxity.facility.Facility
import com.pluxity.global.constant.ErrorCode.NOT_FOUND_FACILITY_PATH
import com.pluxity.global.exception.CustomException
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class FacilityPathService(
    private val facilityPathRepository: FacilityPathRepository,
) {
    @Transactional(readOnly = true)
    fun findById(id: Long): FacilityPath =
        facilityPathRepository.findByIdOrNull(id)
            ?: throw CustomException(NOT_FOUND_FACILITY_PATH, id)

    @Transactional
    fun update(
        pathId: Long,
        name: String?,
        type: String?,
        path: String?,
    ) {
        val facilityPath = findById(pathId)
        name?.takeIf { it.isNotBlank() }?.let { facilityPath.updateName(it) }
        type?.takeIf { it.isNotBlank() }?.let { facilityPath.updatePathType(PathType.from(it)) }
        path?.takeIf { it.isNotBlank() }?.let { facilityPath.updatePath(it) }
    }

    @Transactional
    fun save(
        facility: Facility,
        name: String,
        type: String,
        path: String,
    ) {
        facilityPathRepository.save(
            FacilityPath(
                facility = facility,
                name = name,
                pathType = PathType.from(type),
                path = path,
            ),
        )
    }

    @Transactional
    fun delete(pathId: Long) {
        facilityPathRepository.delete(findById(pathId))
    }
}
